clash-cores/src/Clash/Cores/Xilinx/DcFifo/Explicit.hs

```diff
@@ -0,0 +1,247 @@
+{-|
+Copyright  :  (C) 2022, Google Inc,
+License    :  BSD2 (see the file LICENSE)
+Maintainer :  QBayLogic B.V. <devops@qbaylogic.com>
+
+Support for the [FIFO Generator v13.2](https://docs.xilinx.com/v/u/en-US/pg057-fifo-generator).
+
+
+-}
+
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE QuasiQuotes #-}
+
+{-# OPTIONS_GHC -Wno-unrecognised-pragmas #-}
+
+module Clash.Cores.Xilinx.DcFifo.Explicit
+  ( -- * Instantiating IP
+    dcFifo
+  , XilinxFifo(..)
+
+    -- * Customizing IP
+  , DcConfig(..)
+  , DcImplementation(..)
+  , ReadMode(..)
+  , defConfig
+
+    -- * Helper type aliases
+  , ResetBusy
+  , Full
+  , Empty
+  , DataCount
+  ) where
+
+import           Clash.Annotations.Primitive (Primitive (InlineYamlPrimitive))
+import           Clash.Explicit.Prelude
+import           Clash.Signal.Internal       (Signal (..))
+
+
+import qualified Data.Sequence               as Seq
+import           Data.String.Interpolate     (i)
+
+-- We want constructor names to correspond 1:1 to Xilinx options it ease
+-- blackbox implementation.
+{-# HLINT ignore "Use camelCase" #-}
+
+type ResetBusy = Bool
+type Full = Bool
+type Empty = Bool
+type DataCount n = Unsigned n
+
+data ReadMode
+  = Standard_FIFO
+  | First_Word_Fall_Through
+  deriving (Show, Generic)
+
+data DcImplementation
+  = Independent_Clocks_Block_RAM
+  | Independent_Clocks_Distributed_RAM
+  | Independent_Clocks_Builtin_FIFO
+  deriving (Show, Generic)
+
+data DcConfig depth = DcConfig
+  { dcDepth          :: !depth
+  , dcImplementation :: !DcImplementation
+  , dcReadMode       :: !ReadMode
+  , dcReadDataCount  :: !Bool
+  , dcWriteDataCount :: !Bool
+  , dcOverflow       :: !Bool
+  , dcUnderflow      :: !Bool
+  }
+  deriving (Show, Generic)
+
+defConfig :: KnownNat depth => DcConfig (SNat depth)
+defConfig = DcConfig
+  { dcDepth = SNat
+  , dcImplementation = Independent_Clocks_Block_RAM
+  , dcReadMode = Standard_FIFO
+  , dcReadDataCount = True
+  , dcWriteDataCount = True
+  , dcOverflow = False
+  , dcUnderflow = False
+  }
+
+data FifoState n = FifoState
+  { hsQueue      :: Seq.Seq (BitVector n)
+  , relativeTime :: Int -- ^ In order to model a FIFO in two clock domains,
+                        -- we track relative offset in order to know which
+                        -- signals to peel off.
+  } deriving Show
+
+data XilinxFifo read write depth n =
+  XilinxFifo
+    { writeReset :: Signal write ResetBusy -- ^ @wr_rst@
+    , isFull :: Signal write Full -- ^ @full@
+    , isOverflow :: Signal write Bool -- ^ @overflow@
+    , writeCount :: Signal write (DataCount depth) -- ^ @wr_data_count@
+    , readReset :: Signal read ResetBusy -- ^ @rd_rst@
+    , isEmpty :: Signal read Empty -- ^ @empty@
+    , isUnderflow :: Signal read Bool -- ^ @underflow@
+    , readCount :: Signal read (DataCount depth) -- ^ @rd_data_count@
+    , fifoData :: Signal read (BitVector n) -- ^ @dout@
```

Comment:
I have found a problem I think - the `rd_rst_busy` signal is available only for built-in FIFOs. Which is not what we use!

Reply:
Erm you mean they are only available for BlockRAM implementations, right?

Reply:
From the data sheet:

![Screenshot from 2022-07-14 14-26-41](https://user-images.githubusercontent.com/5658854/178981989-ae12382e-fd47-4c57-bbc8-f95d766df5af.png)

Reply:
On `wr_rst_busy`:

![Screenshot from 2022-07-14 14-27-42](https://user-images.githubusercontent.com/5658854/178982231-64bb0fab-a2c6-43ee-8014-c73c4e69a408.png)

Reply:
Welp, I grabbed the wrong sections. Here are  the docs for FIFOs with independent clocks:

![Screenshot from 2022-07-14 14-29-58](https://user-images.githubusercontent.com/5658854/178982681-59436a26-b721-4b1b-8d49-6bc997cd00de.png)

and

![Screenshot from 2022-07-14 14-29-41](https://user-images.githubusercontent.com/5658854/178982682-9b1ec437-64c4-4c20-8f49-86ed1e4051ef.png)

Reply:
![image](https://user-images.githubusercontent.com/3692925/178982747-4d188f5b-ccbd-4022-b95b-00805c590a03.png)
I don't have any UltraScale devices in my Vivado so I can't see that option, but the wizard is happy to put it on Independent Clock Block RAM. Who is right, the datasheet or the wizard?

Reply:
> Yeah we're not using common clock, right? We are using dual-clock.

Yep, sorry, I'll delete the other posts ~out of shame~ to streamline the discussion.

Reply:
It wouldn't feel good relying on something the spec says is unsupported, IMO. Maybe:

 * We can leave out `rd_rst_busy`. During reset `empty` will be asserted anyway, so there's no risk of submitting spurious read commands.
 * We could add `wr_ack` to test whether a write has been accepted.

I do wonder though; maybe the datasheet states that a FIFO will be ready to accept writes _immediately_ after releasing the reset? (Provided that the reset is asserted for long enough.)

Reply:
No, first of all the reset is asynchronous (necessarily since it is an input to both clock domains); it is synchronised by the FIFO core. So deasserting the reset input will not give you any guarantees about readiness of the write side. But if they assert `full` while in reset, no extra logic is necessary.

Reply:
Could we ask in the Xilinx forums?

Reply:
I'll simply have a look whether `full` is asserted in simulation.

Reply:
I'll just note a few things that are no longer an issue now that we only do `Independent_Clocks_Block_RAM`:

* Not all variants have `wr_rst_busy` and `rd_rst_busy`, even depending on the chip family (UltraScale or not).
* The UltraScale variant of builtin FIFO has no input named `rst`; it has an `srst` synchronous to the write clock.
* Buitin FIFO requires you to specify the properties `CONFIG.Read_Clock_Frequency` and `CONFIG.Write_Clock_Frequency` (values in MHz).
* Builtin FIFO does not support data counts.

Instantiating a FIFO that has no `..._rst_busy` signals does not simulate:
```
ERROR: [VRFC 10-718] formal port <wr_rst_busy> does not exist in entity <dcfifo>.  Please compare the definition of block <dcfifo> to its component declaration and its instantion to detect the mismatch. [/tmp/clash-test-d6df6e43cacf34db/DcFifo0.topEntity/topEntity.vhdl:67]
```

When I edit one of the tests in the test bench to instantiate a `Independent_Clocks_Builtin_FIFO`, the test bench fails.

Reply:
> I'll simply have a look whether full is asserted in simulation.

Yes, it is asserted during the reset phase. I don't think we need the `busy` signals.

Reply:
Ok, I disabled the `reset_busy` signals.

Reply:
I think we should have a test for the behaviour of `full` and `empty` during the reset phase, but it isn't trivial at all. I think I did come up with a fair test, but I'd like to make an issue for it that we need to add such a test rather than hold up this PR.